run/local-cluster.go

```diff
@@ -0,0 +1,402 @@
+
+// Run using: go run local_cluster.go
+
+// +build ignore
+
+package main
+
+import (
+	"crypto/tls"
+	"encoding/json"
+	"flag"
+	"fmt"
+	"io/ioutil"
+	"log"
+	"net"
+	"net/http"
+	"net/url"
+	"os"
+	"os/signal"
+	"strings"
+	"time"
+
+	"github.com/fsouza/go-dockerclient"
+)
+
+const (
+	dockerspyImage = "iverberk/docker-spy"
+	domain         = "local"
+)
+
+var cockroachImage = flag.String("i", "cockroachdb/cockroach-dev", "the docker image to run")
+var cockroachEntry = flag.String("e", "", "the entry point for the image")
+var numNodes = flag.Int("n", 3, "the number of nodes to start")
+var waitInterrupt = flag.Bool("w", false, "wait for an interrupt before existing")
+
+func prettyJSON(v interface{}) string {
+	pretty, err := json.MarshalIndent(v, "", "  ")
+	if err != nil {
+		panic(err)
+	}
+	return string(pretty)
+}
+
+func newDockerClient() *docker.Client {
+	client, err := docker.NewTLSClient(
+		os.ExpandEnv("${DOCKER_HOST}"),
+		os.ExpandEnv("${DOCKER_CERT_PATH}/cert.pem"),
+		os.ExpandEnv("${DOCKER_CERT_PATH}/key.pem"),
+		os.ExpandEnv("${DOCKER_CERT_PATH}/ca.pem"))
+	if err != nil {
+		log.Fatal(err)
+	}
+	return client
+}
+
+// remove a container (docker rm)
+func remove(client *docker.Client, c *docker.Container) {
+	err := client.RemoveContainer(docker.RemoveContainerOptions{
+		ID:            c.ID,
+		RemoveVolumes: true,
+	})
+	if err != nil {
+		panic(err)
+	}
+}
+
+// kill and remove a container (docker kill && docker rm)
+func kill(client *docker.Client, c *docker.Container) {
+	err := client.KillContainer(docker.KillContainerOptions{
+		ID: c.ID,
+	})
+	if err != nil {
+		panic(err)
+	}
+	remove(client, c)
+}
+
+// start a container (docker start)
+func start(client *docker.Client, dns, c *docker.Container) {
+	var dnsAddrs []string
+	if dns != nil {
+		dnsAddrs = append(dnsAddrs, dns.NetworkSettings.IPAddress)
+	}
+	err := client.StartContainer(c.ID, &docker.HostConfig{
+		PublishAllPorts: true,
+		DNS:             dnsAddrs,
+	})
+	if err != nil {
+		panic(err)
+	}
+}
+
+// wait for a container to exit (docker wait)
+func wait(client *docker.Client, c *docker.Container) {
+	code, err := client.WaitContainer(c.ID)
+	if err != nil {
+		panic(err)
+	}
+	if code != 0 {
+		logs(client, c)
+		panic(fmt.Errorf("non-zero exit code: %d", code))
+	}
+}
+
+// retrieve the logs for a container (docker logs)
+func logs(client *docker.Client, c *docker.Container) {
+	err := client.Logs(docker.LogsOptions{
+		Container:    c.ID,
+		OutputStream: os.Stdout,
+		ErrorStream:  os.Stderr,
+		Stdout:       true,
+		Stderr:       true,
+	})
+	if err != nil {
+		panic(err)
+	}
+}
+
+// inspect a container (docker inspect)
+func inspect(client *docker.Client, cid string) *docker.Container {
+	c, err := client.InspectContainer(cid)
+	if err != nil {
+		panic(err)
+	}
+	return c
+}
+
+// retrieve the ip address of docker itself
+func dockerIP() string {
+	u, err := url.Parse(os.ExpandEnv("${DOCKER_HOST}"))
```

Comment:
Use os.Getenv instead of os.ExpandEnv when the variable expansion is the only thing in the string.